File: 03-Stacks/java/src/Main.java
import java.awt.*;
import java.util.Stack;

public class Main {
    public static void main(String[] args) {
//        Stack<Character> stack = new Stack<>();

//        stack.push(10);
//        stack.push(20);
//        stack.push(30);
//
//        System.out.println(stack);
//        var top = stack.pop();
//        System.out.println(top);
//
//        top = stack.peek();
//        System.out.println(stack);
//        System.out.println(top);

//        Reversing a String
//        String str = "abcd";
//
//        StringReverser reverser = new StringReverser();
//        System.out.println(reverser.reverse(str));

//        Balanced expression
       /* String input = "[{3*(2-3)}]()";
        Expression exp = new Expression(input);
        System.out.println(exp.isBalanced()); */

//        My Stack
      /*  myStack st = new myStack();

        st.push(30);
        st.push(10);
        st.push(40);
        st.push(90);
        st.pop();

        System.out.println(st.peek());*/

//        TwoStacks
        System.out.println("TwoStacks");

        TwoStacks twoStacks = new TwoStacks(4);

        twoStacks.push1(4);
        twoStacks.push1(8);
        twoStacks.push2(44);
        twoStacks.push2(54);

        System.out.println(twoStacks);
        twoStacks.pop1();
        twoStacks.pop2();
        System.out.println(twoStacks);

        System.out.println("--------");

//        MinStack
        System.out.println("MinStack");

        MinStack test = new MinStack();

        test.push(5);
        test.push(2);
        test.push(10);
        test.push(1);


        System.out.println(test.min());
        test.pop();
        System.out.println(test.min());

        System.out.println("--------");
        

    }
}

File: 03-Stacks/cpp/main.cpp
#include <iostream>
#include <string>
#include "StringReverser.h"
#include "Expression.h"
#include "myStack.h"
#include "TwoStacks.h"
#include "MinStack.h"

using namespace std;

int main() {

    // Reversing a string
    cout << "Reversing a string" << endl;

    StringReverser reverser;
    string res = reverser.reverse("bcdedit");
    cout << res << endl;

    cout << "--------" << endl;

    // Paren-Matching
    cout << "Paren-Matching" << endl;

    Expression exp("{3*(2+4)}[0]");
    cout << boolalpha << exp.isBalanced() << endl;

    cout << "--------" << endl;


    // myStack
    cout << "myStack" << endl;

    myStack st(5);
    st.push(30);
    st.push(20);
    st.push(40);
    st.push(35);
    st.push(30);
    st.pop();
    cout << st.peek() << endl;

    cout << "--------" << endl;


    // TwoStacks
    cout << "TwoStacks" << endl;

    TwoStacks twoStacks(5);
    twoStacks.push1(4);
    twoStacks.push1(10);

    twoStacks.push2(20);
    twoStacks.push2(40);
    //twoStacks.push2(50);

    while (!twoStacks.isEmpty1() || !twoStacks.isEmpty2()) {

        cout << twoStacks.pop1() << endl;
        cout << twoStacks.pop2() << endl;
    }

    cout << "--------" << endl;


    // MinStack
    cout << "MinStack" << endl;

    MinStack minStack;

    minStack.push(5);
    minStack.push(2);
    minStack.push(10);
    minStack.push(1);

    cout << minStack.min() << endl;
    minStack.pop();
    cout << minStack.min() << endl;

    cout << "--------" << endl;

    return 0;
}
